### PR TITLE
Trim string values defined in the Hibernate ORM configuration

### LINKED ISSUE #20492: Class PostgreSQL10Dialect not found on Quarkus in Google App Engine java11?
### Describe the bug

I wanted to deploy my quarkus application including hibernate panache to GCP app engine standard and got the error: Class PostgreSQL10Dialect not found on Quarkus

### Expected behavior

Successful startup of quarkus on GCP app engine standard

### Actual behavior

I've set up my quarkus application regarding to https://quarkus.io/guides/deploying-to-google-cloud 
Created a Cloud SQL instance.
And happily deployed the java11 app the first time to GAE. :-)

But - on startup quarkus fails with this reason:
```
	Caused by: java.lang.ClassNotFoundException: org.hibernate.dialect.PostgreSQL10Dialect 
	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:581)
	at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:178)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:522)
	at java.base/java.lang.Class.forName0(Native Method)
	at java.base/java.lang.Class.forName(Class.java:398)
	at io.quarkus.hibernate.orm.runtime.service.FlatClassLoaderService.classForName(FlatClassLoaderService.java:36)
```

Looking at the jar:
```
$ jar tvf foodie-1.0-SNAPSHOT-runner.jar | fgrep PostgreSQL10Dialect
  1184 Tue Sep 28 20:12:07 CEST 2021 io/quarkus/hibernate/orm/runtime/dialect/QuarkusPostgreSQL10Dialect.class
   990 Tue Sep 28 20:12:08 CEST 2021 org/hibernate/dialect/PostgreSQL10Dialect.class
```

Any hints how to proceed?

application.properties:
```
quarkus.package.type=uber-jar
quarkus.datasource.db-kind=other 
%dev.quarkus.datasource.jdbc.url=jdbc:postgresql:///quarkus
quarkus.datasource.jdbc.url=jdbc:postgresql:///quarkus
quarkus.datasource.jdbc.driver=org.postgresql.Driver
quarkus.datasource.jdbc.additional-jdbc-properties.cloudSqlInstance=addlogic-foodiefnf-1:europe-west3:quarkus 
quarkus.datasource.jdbc.additional-jdbc-properties.socketFactory=com.google.cloud.sql.postgres.SocketFactory
quarkus.hibernate-orm.dialect=org.hibernate.dialect.PostgreSQL10Dialect 
```

Extract from pom.xml:
```
		<maven.compiler.target>11</maven.compiler.target>
		<quarkus-plugin.version>2.2.3.Final</quarkus-plugin.version>
		<quarkus.platform.artifact-id>quarkus-universe-bom</quarkus.platform.artifact-id>
		<quarkus.platform.group-id>io.quarkus</quarkus.platform.group-id>
		<quarkus.platform.version>2.2.3.Final</quarkus.platform.version>
		<postgres-socket-factory.version>1.3.3</postgres-socket-factory.version>
[...]
		<dependency>
			<groupId>io.quarkus</groupId>
			<artifactId>quarkus-jdbc-postgresql</artifactId>
		</dependency>
		<dependency>
			<groupId>io.quarkus</groupId>
			<artifactId>quarkus-hibernate-orm-panache</artifactId>
		</dependency>
		<dependency>
			<groupId>com.google.cloud.sql</groupId>
			<artifactId>postgres-socket-factory</artifactId>
			<version>${postgres-socket-factory.version}</version>
		</dependency>
```

Sidenote on deployment?
----------------------------
How to investigate this warning on ´mvn clean package appengine:deploy´?

```
[INFO] Staging the application to: D:\eclipse\eclipse-workspace\code-with-quarkus\target\appengine-staging
[INFO] Detected App Engine app.yaml based application.
Okt. 01, 2021 8:08:38 AM com.google.cloud.tools.appengine.operations.AppYamlProjectStaging copyArtifactJarClasspath
WARNING: Could not copy 'Class-Path' jar: D:\eclipse\eclipse-workspace\code-with-quarkus\target referenced in MANIFEST.MF
```

Just a sidenote as deployment with `gcloud app deploy target/code-with-quarkus-1.0.0-SNAPSHOT-runner.jar` gives no WARNINFS and has same effects.

Alternative with mysql is functioning
=====================================
I've tried an alternative with mysql, which is functioning in the same MCVE and another testing project which I have set up.

pom.xml
```
		<dependency>
			<groupId>io.quarkus</groupId>
			<artifactId>quarkus-jdbc-mysql</artifactId>
		</dependency>
		<dependency>
			<groupId>com.google.cloud.sql</groupId>
			<artifactId>mysql-socket-factory-connector-j-8</artifactId>
			<version>1.3.3</version>
		</dependency>
```

application.properties
```
# https://github.com/GoogleCloudPlatform/cloud-sql-jdbc-socket-factory/blob/main/docs/jdbc-mysql.md
quarkus.datasource.db-kind=other 
quarkus.datasource.jdbc.url=jdbc:mysql:///quarkus 
quarkus.datasource.jdbc.driver=com.mysql.cj.jdbc.Driver
quarkus.datasource.username=root
quarkus.datasource.password=quarkus
quarkus.datasource.jdbc.additional-jdbc-properties.cloudSqlInstance=addlogic-foodiefnf-1:europe-west3:quarkusmysql
quarkus.datasource.jdbc.additional-jdbc-properties.socketFactory=com.google.cloud.sql.mysql.SocketFactory
quarkus.hibernate-orm.dialect=org.hibernate.dialect.MySQL8Dialect
```

***So this makes me more curious why it is not working with PostgreSQL.***

### How to Reproduce?

An MCVE reproducer project can be found at 
`git@gitlab.com:addlogic/code-with-quarkus.git#`

See README for reproduction steps.

### Output of `uname -a` or `ver`

CYGWIN_NT-10.0 levante2b 3.2.0(0.340/5/3) 2021-03-29 08:42 x86_64 Cygwin

### Output of `java -version`

$ java -version java version "11.0.11" 2021-04-20 LTS Java(TM) SE Runtime Environment 18.9 (build 11.0.11+9-LTS-194) Java HotSpot(TM) 64-Bit Server VM 18.9 (build 11.0.11+9-LTS-194, mixed mode)

### GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

2.2.3.Final

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

apache-maven-3.8.1

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by gsmet**
Fixes #20492

### CODE REVIEW

File: core/runtime/src/main/java/io/quarkus/runtime/configuration/CharsetConverter.java [hunk 1]
```diff
@@ -19,10 +19,16 @@
 
     @Override
     public Charset convert(String value) {
+        if (value == null) {
+            return null;
+        }
+
+        String trimmedCharset = value.trim();
```
Comment (yrodiere) [inline core/runtime/src/main/java/io/quarkus/runtime/configuration/CharsetConverter.java:26]:
```suggestion
        String trimmedCharset = value.trim();

        if (trimmedCharset.isEmpty()) {
            return null;
        }
```

**Review by yrodiere — APPROVED** [review]
+1 to merge this, but... why in the world is trimming not the default?

I suppose it's something related to the Microprofile spec, or an ongoing battle with someone else arguing we should do no such thing?

I cannot think of any property (at least in Hibernate ORM/Search) where we would assign meaning to a leading/trailing space. And even if there was one, it would be much safer to add an exception for that property than to manually use `TrimmedStringConverter` on every single property except the one that should not be trimmed...

Hibernate Search, internally, certainly does trim everything before processing any configuration property, but any processing in the Hibernate Search (e.g. processing a bean reference extracted from the configuration properties) is at risk too. So I guess I'll have to do the same as you for the Hibernate Search extension... ?

**Comment by gsmet** [comment]
That would be a question for @radcortez . I must admit that I agree with you. We had a discussion about this back in the days with David and in the end the initial behavior was kept. IIRC the main argument for it was to respect how .properties work and their trimming rules.

I could see how you would want to not trim user-defined properties though (think email subject prefix for instance). But I agree that would be marginal.

That being said, I'm not sure that's something we can break now.

**Comment by radcortez** [comment]
> I suppose it's something related to the Microprofile spec, or an ongoing battle with someone else arguing we should do no such thing?

No :)

> IIRC the main argument for it was to respect how .properties work and their trimming rules.

Exactly. 

Maybe what we can do is trim properties in the `quarkus` namespace and leave the rest.

**Comment by yrodiere** [comment]
> Maybe what we can do is trim properties in the `quarkus` namespace and leave the rest.

That would be awesome. We can always override the converter if we don't need trimming for one exotic property.

**Comment by radcortez** [comment]
> > Maybe what we can do is trim properties in the `quarkus` namespace and leave the rest.
> 
> That would be awesome. We can always override the converter if we don't need trimming for one exotic property.

Exactly. @gsmet do you remember any config on our side where we may want to conserve the original value?

**Comment by gsmet** [comment]
> do you remember any config on our side where we may want to conserve the original value?

That's a good question, thanks for asking... :). I think the only way to really know that would be to do a quick scan of all the configuration properties in the configuration reference page. But I can't do that anymore with my bad eyes :/.

I'm not sure having a different behavior for a specific namespace is a good idea, though.

In any case:
- we need to discuss this on the dev list, not in a somewhat unrelated PR
- we won't do that for 2.7, it's 2.8 material so I think we should get this PR in anyway

**Review by radcortez — APPROVED** [review]

**Comment by radcortez** [comment]
Sure. I was not proposing to do it here.

I'll open a separate issue so we don't forget.

**Pull request merged**